### PR TITLE
SMACT Metallicity Handling Enhancements

### PR DESCRIPTION
**Pull request by ryannduma**
# SMACT Metallicity/Alloy Handling Enhancement

This document describes the enhancements made to SMACT to better handle metallic compounds and their classification. The changes introduce a scoring system for identifying and validating metallic compounds, moving beyond simple binary classification to provide a more nuanced understanding of a compound's metallic character.

### 1. Module: `smact.metallicity`

A dedicated module for handling metallic compounds with several specialized functions:

#### `get_metal_fraction(composition)`

- Calculates the fraction of metallic elements in a composition
- Input: Chemical formula (str) or pymatgen Composition object
- Output: Float between 0-1
- Example:

```python
from smact.metallicity import get_metal_fraction

# Pure metallic compound - returns 1.0
print(get_metal_fraction("Fe3Al"))  # Works with string formula (& Composition object)

# Mixed compound - returns fraction of how much of the compound is metallic
print(get_metal_fraction("Fe2O3"))  # 0.4
```

#### `get_d_electron_fraction(composition)`

- Calculates the fraction of d-block elements
- Important for transition metal compounds
- Example:

```python
from smact.metallicity import get_d_electron_fraction

# Pure transition metal compound - returns 1.0
print(get_d_electron_fraction("Fe2Nb"))

# Mixed compound - returns fraction
print(get_d_electron_fraction("Fe3Al"))  # 0.75 (Fe is d-block, Al is not)
```

#### `get_distinct_metal_count(composition)`

- Counts unique metallic elements
- Useful for identifying complex metallic systems
- Example:

```python
from smact.metallicity import get_distinct_metal_count

# Binary metallic compound
print(get_distinct_metal_count("Fe3Al"))  # Returns 2

# Complex multi-component composition
print(get_distinct_metal_count("NbTiAlCr"))  # Returns 4

# Non-metallic compound
print(get_distinct_metal_count("SiO2"))  # Returns 0
```

#### `get_pauling_test_mismatch(composition)`

- Calculates deviation from ideal electronegativity ordering
- Handles both metal-metal and metal-nonmetal pairs
- Lower scores indicate more metallic-like bonding
- Example:

```python
from smact.metallicity import get_pauling_test_mismatch

# Metallic compound - low mismatch
print(get_pauling_test_mismatch("Fe3Al"))  # 0.22

# Ionic compound - high mismatch
print(get_pauling_test_mismatch("NaCl"))  # -1.23
```

#### `metallicity_score(composition)`

- Main scoring function combining multiple metrics
- Returns a score between 0-1
- Higher scores indicate more metallic character
- Example:

```python
from smact.metallicity import metallicity_score

# Metallic compounds - high scores
print(metallicity_score("Fe3Al"))  # ~0.83
print(metallicity_score("Ni3Ti"))  # ~0.87

# Non-metallic compounds - low scores
print(metallicity_score("NaCl"))  # 0.63
print(metallicity_score("Fe2O3"))  # 0.64
print(metallicity_score("SiO2"))  # 0.25
```

### 2. Enhanced `smact_validity`

The existing `smact_validity` function in `smact.screening` has been enhanced with three validation paths:

1. Standard validation (charge neutrality and electronegativity)
2. Simple metal validation
3. Metallicity scoring validation

Example usage showing all validation paths:

```python
from smact.screening import smact_validity

# Test with different validation methods
compound = "Fe3Al"

# 1. Standard validation (no alloy/metallicity check)
is_valid_standard = smact_validity(
    compound, use_pauling_test=True, include_alloys=False, check_metallicity=False
)

# 2. With alloy detection
is_valid_alloy = smact_validity(
    compound, use_pauling_test=True, include_alloys=True, check_metallicity=False
)

# 3. With metallicity detection
is_valid_metallic = smact_validity(
    compound,
    use_pauling_test=True,
    include_alloys=False,
    check_metallicity=True,
    metallicity_threshold=0.7,
)

# Or combine methods
is_valid = smact_validity(
    compound,
    use_pauling_test=True,
    include_alloys=True,
    check_metallicity=True,
    metallicity_threshold=0.7,
)
```

### 3. Comprehensive Analysis Example

Here's how to perform a detailed analysis of a compound:

```python
from smact.metallicity import *
from smact.screening import smact_validity


def analyze_compound(formula):
    """Perform comprehensive analysis of a compound."""
    # Basic metrics
    metal_frac = get_metal_fraction(formula)
    d_frac = get_d_electron_fraction(formula)
    n_metals = get_distinct_metal_count(formula)
    pauling = get_pauling_test_mismatch(formula)
    score = metallicity_score(formula)

    # Validity checks
    valid_standard = smact_validity(
        formula, use_pauling_test=True, include_alloys=False, check_metallicity=False
    )
    valid_alloy = smact_validity(
        formula, use_pauling_test=True, include_alloys=True, check_metallicity=False
    )
    valid_metallic = smact_validity(
        formula, use_pauling_test=True, include_alloys=False, check_metallicity=True
    )

    print(f"Analysis of {formula}:")
    print(f"Metal fraction: {metal_frac:.2f}")
    print(f"d-electron fraction: {d_frac:.2f}")
    print(f"Distinct metals: {n_metals}")
    print(f"Pauling mismatch: {'nan' if math.isnan(pauling) else f'{pauling:.2f}'}")
    print(f"Metallicity score: {score:.2f}")
    print(f"Valid (standard): {valid_standard}")
    print(f"Valid (alloy): {valid_alloy}")
    print(f"Valid (metallic): {valid_metallic}")


# Example usage
compounds = [
    "Fe3Al",  # Classic intermetallic
    "NaCl",  # Ionic
    "Fe2O3",  # Metal oxide
    "Cu2MgSn",  # Heusler alloy
    "NbTiAlCr",  # High-entropy alloy
]

for compound in compounds:
    analyze_compound(compound)
    print("-" * 50)
```

## Future Development Directions

1. **Specialized Submodules**
   - Development of specific modules for different types of metallic systems:
     - Intermetallic compound prediction
     - Solid solution formation
     - Amorphous metal formation
   - Integration with structure prediction

   - Incorporate atomic size factors
   - Add structure prediction capabilities
   - Include formation energy estimates - linking back to Miedema's model [DOI Link](https://doi.org/10.1016/j.cpc.2016.08.013)

3. **Validation and Refinement**
   - Benchmark against experimental databases
   - Refine scoring weights with more data
   - Add support for mixed-valence compounds

3. **Extended Functionality**
   - Add support for partial occupancies
   - Include temperature-dependent properties
   - Integrate with phase diagram tools

## Contributing

Contributions to improve the metallicity functionality are welcome! Areas particularly in need of development:

1. Additional test cases and validation
2. Refinement of scoring weights
3. Integration with external databases
4. Development of specialized submodules for specific metallic systems
5. Performance optimization for large-scale screening

## References

1. Original SMACT paper: [SMACT: Semiconducting Materials by Analogy and Chemical Theory](https://joss.theoj.org/papers/10.21105/joss.01361)
2. Intermetallics theory and classification literature sources:

   - D.G. Pettifor introduced the concept of a single "chemical scale" or "structure map" coordinate (Pettifor number) to systematically separate compound classes [1, p. 31]. The new intermetallicscore is a step in that direction but customized to SMACT's internal data structures.

     - Reference: D.G. Pettifor, "A chemical scale for crystal-structure maps," Solid State Communications. 51 (1984) 31–34. [DOI Link](<https://doi.org/10.1016/0038-1098(84)90765-8>)

   - The role of charge transfer and atomic size mismatch is pivotal in stabilizing intermetallic phases. Miedema's framework quantifies these effects, making it useful for predicting alloying behaviors and crystal structure. The parameters coded here, while conceptually similar, have not implemented Miedema's model directly.
   - Reference: A.R. Miedema, Cohesion in alloys - fundamentals of a semi-empirical model.[DOI Link](<https://doi.org/10.1016/0378-4363(80)90054-6>)

3. Electronegativity scales and their role in predicting bonding character (Pauling electronegativity)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test A - Unit Tests
- [x] Test B - XGBoost Classification exercise

### Test A - Unit Tests (smact/tests/test_intermetallics.py):
- Tests individual feature functions
- Validates edge cases
- Key test cases include
```python
  # Known intermetallics
  self.intermetallics = [
      "Fe3Al",  # Classic intermetallic
      "Ni3Ti",  # Superalloy component
      "Cu3Au",  # Ordered alloy
      "Fe2Nb",  # Laves phase
  ]

  # Known non-intermetallics
  self.non_intermetallics = [
      "NaCl",   # Ionic
      "SiO2",   # Covalent
      "Fe2O3",  # Metal oxide
      "CuSO4",  # Complex ionic
  ]
```

### Test B: Machine Learning Classification

The metallicity module has been used in conjunction with machine learning approaches to predict the metallic nature of compounds. The `metallicity_classification.ipynb` notebook demonstrates how these chemical descriptors can be used to train models for predicting whether a compound will exhibit metallic behavior.

The classification task uses features derived from the metallicity module including:
- Metal fraction
- d-electron fraction
- Number of distinct metals
- Pauling electronegativity mismatch
- Overall metallicity score

These features are used to train models that can predict the likelihood of metallic behavior in compounds, which is more general than specifically predicting intermetallic behavior. This approach allows for a broader understanding of metallic character in materials, encompassing various types of metallic systems including:
- Traditional metals and alloys
- Intermetallic compounds
- Solid solutions
- Metallic glasses
- Mixed-character compounds
- Uses matbench_expt_is_metal dataset
- Implements XGBoost classifier
- Performs feature importance analysis

Classification Workflow Details:

1. Data Loading & Preparation:
```python 
from matminer.datasets import load_dataset
df = load_dataset("matbench_expt_is_metal")  # 4921 entries
```

2. Feature Extraction:

- Metal fraction
- d-electron fraction
- Distinct metal count
- Pauling mismatch
- Overall intermetallic score

3. Model Training:

- XGBoost classifier with hyperparameter tuning
- Cross-validation with threshold optimization
- Feature importance analysis using SHAP

### Classification results:

```bash
Test Accuracy: 0.858 with threshold=0.5

Classification Report:
              precision    recall  f1-score   support

       False       0.84      0.87      0.85       494
        True       0.86      0.83      0.85       491

    accuracy                           0.85       985
   macro avg       0.85      0.85      0.85       985
weighted avg       0.85      0.85      0.85       985
``` 
confusion matrix

Reproduction Instructions: 
1. Environment Setup:
```python
# Create conda environment
conda create -n smact-env python=3.11
conda activate smact-env

# Install dependencies
pip install smact matminer xgboost shap scikit-learn pandas numpy matplotlib seaborn
```
2. Run Unit Tests
```python 
python -m pytest smact/tests/test_intermetallics.py -v
``` 

3. Run Classification Notebook:
```python
jupyter notebook metallicity_classification.ipynb
```
### Key Results

Unit Test Results:

- All feature functions pass edge cases
- Intermetallic scoring correctly differentiates between known compounds
- Error handling works as expected

2. Classification Performance:
- High accuracy in metal vs. non-metal classification
- Feature importance ranking shows:
  1. Metal fraction
  2. d-electron fraction
  3. Intermetallic score
  4. Pauling mismatch
  5. Distinct metal count

**Test Configuration**:

- Python version: 3.11
- Operating System: Linux 
- Key package versions: 
  - smact: latest
  - pymatgen: latest
  - xgboost: latest
  - scikit-learn: latest
  - pandas: latest
  - numpy: latest
  - shap: latest

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

## Release Notes

- **New Features**
  - Introduced a metallicity scoring system for chemical compounds.
  - Added advanced validation for metallic compounds.
  - Enhanced screening functionality with metallicity checks.
  - New utility functions for analysing metallic characteristics.
  - Comprehensive tutorial on metallicity classification added.

- **Documentation**
  - Expanded documentation for the metallicity module.
  - Included new examples related to metallicity in the documentation.
  - Updated tutorials to include metallicity classification.

- **Dependencies**
  - Introduced machine learning-related optional dependencies.
  - Updated project requirements to support new functionality.

The update provides more sophisticated tools for analysing and validating metallic compounds, with improved scoring mechanisms and comprehensive documentation.

### CODE REVIEW

File: docs/intermetallics_readme.md [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+```python
+from smact.intermetallics import intermetallic_score
+
+# Classic intermetallics - high scores
+print(intermetallic_score("Fe3Al"))  # ~0.85
+print(intermetallic_score("Ni3Ti"))  # ~0.82
+
+# Non-intermetallics - low scores
+print(intermetallic_score("NaCl"))  # ~0.20
+print(intermetallic_score("Fe2O3"))  # ~0.45
```
Comment (AntObi) [inline docs/intermetallics_readme.md:94]: Can you verify these numbers?
Running these lines gives me very different numbers
image

**Review by AntObi — DISMISSED** [review]

File: smact/intermetallics.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    electronegativities = [el.pauling_eneg for el in elements]
+
+    # Skip if any electronegativities are None
+    if None in electronegativities:
+        return 0.0
+
+    # Calculate pairwise differences
+    mismatches = []
+    for i, (el1, eneg1) in enumerate(zip(elements, electronegativities, strict=False)):
+        for el2, eneg2 in zip(elements[i + 1 :], electronegativities[i + 1 :], strict=False):
```
[the remark this answers is not in the document]
Comment (ryannduma) [reply smact/intermetallics.py:108]: @AntObi when I attempt to make this change then re-run pre-commit, it changes it back to having the `strict=False` argument. I don't know if there's a workaround, cause any attempts to commit break ruff linting

**Comment by ryannduma** [comment]
Hey Anthony, thanks for all the reviews and requested changes, hopefully resolved by the changes made today. Tests are passing which is great, only test not working is pre-commit because there's a new Pyright version out for the pre-commit hook. Other than that, I hope this is a step in the right direction for this chemical filter and I look forward to future discussions and contributions towards its enhancement. Have a lovely weekend and a great start at your new job next week! All the very best <3

**Comment by ryannduma** [comment]
Hey Anthony, latest commit just moved the Jupyter notebook example into the examples docs folder, there are no inherent code changes, though the tests are failing cause of something to do with mp_api being none which I think is a separate issue you're working on so don't worry.

File: smact/metallicity.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+import numpy as np
+from pymatgen.core import Composition
+
+import smact
+from smact import Element
+from smact.properties import valence_electron_count
+
+
+def _ensure_composition(composition: str | Composition) -> Composition:
```
[the remark this answers is not in the document]
Comment (ryannduma) [reply smact/metallicity.py:13]: i think this might be a redundancy on my end, but I originally wrote this to ensure the composition was a pymatgen composition object even in contexts where they may enter a composition that's otherwise - ill do some testing on my end and possibly get rid of the function should this be the case and incorporate your suggested changes.

File: smact/metallicity.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+def get_metal_fraction(composition: str | Composition) -> float:
+    """Calculate the fraction of metallic elements in a composition.
+
+    Implemented using get_element_fraction helper with smact.metals set.
+    """
+    return get_element_fraction(composition, smact.metals)
+
+
+def get_d_electron_fraction(composition: str | Composition) -> float:
```
[the remark this answers is not in the document]
Comment (ryannduma) [reply smact/metallicity.py:64]: Gotcha! I'll fix this in my next commit

**Review by AntObi — APPROVED** [review]
LGTM!
